### PR TITLE
Raise an event when the page is really going to be unloaded (#667). Part 1  - location.

### PR DESCRIPTION
**Pull request by churkin**
\cc @VasilyStrelyaev @AlexanderMoskovkin @LavrovArtem

### CODE REVIEW

File: src/client/redirect-catcher.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export default class RedirectCatcher extends EventEmiter {
+    constructor (codeInstrumentation) {
+        super();
+
+        this.CATCHED_EVENT = 'hammerhead|event|redirect-catched';
+
+        var locationAccessorsInstrumentation = codeInstrumentation.locationAccessorsInstrumentation;
+        var propertyAccessorsInstrumentation = codeInstrumentation.propertyAccessorsInstrumentation;
+        var HASH_RE                          = /#.*$/;
```
Comment (AlexanderMoskovkin) [inline src/client/redirect-catcher.js]: I suppose it should be `const HASH_RE`? And maybe place it before the class definition, as we usually do with consts

File: src/client/sandbox/code-instrumentation/location/wrapper.js [hunk 2]
```diff
@@ -66,23 +78,51 @@ export default class LocationWrapper {
             }
         }));
 
+        var overridePorperty = porperty => {
```
Comment (AlexanderMoskovkin) [inline src/client/sandbox/code-instrumentation/location/wrapper.js]: typo: `porperty `

File: src/client/redirect-catcher.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        var lastLocationValue                = window.location.toString();
+        var locationChangedHandler           = newLocation => {
+            var currentLocation = lastLocationValue;
+
+            lastLocationValue = window.location.toString();
+
+            if (newLocation !== currentLocation &&
```
Comment (LavrovArtem) [inline src/client/redirect-catcher.js]: May be
```javascript
if (newLocation.replace(HASH_RE, '') !== currentLocation.replace(HASH_RE, ''))
    this.emit(this.CATCHED_EVENT, parseProxyUrl(newLocation).destUrl);
```

Comment (churkin) [inline src/client/redirect-catcher.js]: No, `reload` method does not change `location` value

File: src/client/index.js [hunk 4]
```diff
@@ -42,7 +44,8 @@ class Hammerhead {
             xhrCompleted:       this.sandbox.xhr.XHR_COMPLETED_EVENT,
             xhrError:           this.sandbox.xhr.XHR_ERROR_EVENT,
             xhrSend:            this.sandbox.xhr.XHR_SEND_EVENT,
-            fetchSend:          this.sandbox.fetch.FETCH_REQUEST_SEND_EVENT
+            fetchSend:          this.sandbox.fetch.FETCH_REQUEST_SEND_EVENT,
+            redirectCatched:    this.redirectCatcher.CATCHED_EVENT
```
Comment (VasilyStrelyaev) [inline src/client/index.js]: redirectDetected

File: src/client/redirect-catcher.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import EventEmiter from './utils/event-emitter';
+import { parseProxyUrl } from '../utils/url';
+
+const HASH_RE = /#.*$/;
+
+export default class RedirectCatcher extends EventEmiter {
```
Comment (VasilyStrelyaev) [inline src/client/redirect-catcher.js]: RedirectWatch

File: src/client/index.js [hunk 6]
```diff
@@ -42,7 +44,8 @@ class Hammerhead {
             xhrCompleted:       this.sandbox.xhr.XHR_COMPLETED_EVENT,
             xhrError:           this.sandbox.xhr.XHR_ERROR_EVENT,
             xhrSend:            this.sandbox.xhr.XHR_SEND_EVENT,
-            fetchSend:          this.sandbox.fetch.FETCH_REQUEST_SEND_EVENT
+            fetchSend:          this.sandbox.fetch.FETCH_REQUEST_SEND_EVENT,
```
Comment (VasilyStrelyaev) [inline src/client/index.js]: fetchSent:          this.sandbox.fetch.FETCH_REQUEST_SENT_EVENT,

File: src/client/index.js [hunk 7]
```diff
@@ -42,7 +44,8 @@ class Hammerhead {
             xhrCompleted:       this.sandbox.xhr.XHR_COMPLETED_EVENT,
             xhrError:           this.sandbox.xhr.XHR_ERROR_EVENT,
             xhrSend:            this.sandbox.xhr.XHR_SEND_EVENT,
```
Comment (VasilyStrelyaev) [inline src/client/index.js]: xhrSent:            this.sandbox.xhr.XHR_SENT_EVENT,

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by churkin** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by LavrovArtem** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Pull request merged**